### PR TITLE
Installation Data

### PR DESCRIPTION
**Pull request by stairman06**
This is the same as mc-cip/spec#7. Previous discussion and commits can be found there.

### CODE REVIEW

File: implementing_installation.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+The `directory` field in this value should be relative to Minecraft installation that the project is being installed to. This field is subject to path rules defined in **installation.md - Specifying Paths**. Launchers should check for rule violations in any field where a directory is passed in order to ensure format-compliant information.
+
+Files should be placed with the same filename and file extension as when downloaded.
+
+## Implementing `versionJsonInstall`
+
+Launchers will likely already have the mechanisms in place to parse Minecraft version JSON files. To install a Version with the `versionJsonInstall` method, first find a file with the `rel` value set to `versionJson`. If a file is not present, an error should be presented to the user. If multiple files are found, the first one should be used.
+
+Launchers can then use the file with the `versionJson` relation and automatically install libraries from URLs specified in the JSON file.
```
Comment (immibis) [inline implementing_installation.md:15]: What format is the file?

Also I think all the metadata should be in the index file - we shouldn't have files that contain more metadata. Don't know the best way to implement that. Maybe `classpathLibrary` with an additional Maven address on the primary file?

File: installation.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+The file with an installer relation value should be executed. Note that this field is very vague and it's recommended to use something with more information such as `versionJsonInstall`.
+
+### `instanceInstall`
+
+This is a modpack/instance and should be treated as such. Required dependencies should be installed to a directory specific to this instance, then files should be extracted from a file with `"rel" "compressedOverrides"` and moved to the directory.
+
+If this is a modpack, mods should be listed as a required dependency if it is possible to do so.
+
+Files in the `compressedOverrides` zip should be in the root. For example:
+`overrides.zip/overrides/overridden_file.txt` shouldn't be used. Instead, `overrides.zip/overridden_file.txt` is better.
```
Comment (immibis) [inline installation.md:60]: it should if you want the file to be in overrides in the instance directory

File: installation.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+### `other`
+
+An unknown installation method.
```
Comment (immibis) [inline installation.md:64]: Is this the same as not specifying one?

File: installation.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+`TODO: what other methods should be added?`
+
+---
+
+## Specifying Paths
+
+Some fields, such as `placeInDirectory`'s `directory` field, requires a path to be supplied. This path by default is relative to the root of the Minecraft Instance that this project is being installed to. If the Project is not able to be installed to a Minecraft Instance for whatever reason, it is up to launchers to supply a default relative path.
```
Comment (immibis) [inline installation.md:72]: by default - what changes the default?

File: installation.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+Paths MUST use forward slashes to indicate nesting. Backslashes (even on Windows systems) CANNOT be used.
+
+Paths MAY start with `./`, however it is not needed. For example, `./mods` is a valid directory path, however `mods` also works and does not have the unnecessary `./`.
```
Comment (immibis) [inline installation.md:76]: Why may they?

File: installation.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+Absolute paths MAY be specified by prefixing the path with an absolute path prefix. This prefix has two valid options:
```
Comment (immibis) [inline installation.md:78]: Why would you use an absolute path? It's also an obvious security hole.

File: installation.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+The following are rules on paths and filenames:
+
+- Paths that contain Windows restricted filenames and characters are not allowed
+  - Restricted characters are `<`, `>`, `:`, `"`, `|`, `?`, `*`
+- Empty strings are not allowed
+  - If you need to specify the current relative path, without using a subdirectory, use `.`
+- Filenames CANNOT contain a slash or backslash
+- Filenames that end in `.` CANNOT be used
+- Filenames that end in a space CANNOT be used
+- New line and carriage return characters CANNOT be used
```
Comment (immibis) [inline installation.md:94]: The word is MUST NOT

File: installation.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+- Most Unicode characters are allowed, but using normal ASCII characters (A-Z, a-z, 0-9) is recommended
+
+  - Restricted Unicode characters include NUL and CR
```
Comment (immibis) [inline installation.md:97]: LF? Unicode spaces? Byte order marks? Unpaired surrogates?

Note that Windows uses UTF-16 paths and Linux uses arbitrary byte strings. The sensible intersection is ASCII excluding Windows's reserved characters.

**Comment by stairman06** [comment]
Closing this in favor of #5 - it's a much better system.

**Pull request closed without merging**